### DISCUSSION TITLE
Need help root causing otel db.client.connections.usage metric underflow

### DISCUSSION THREAD

**Discussion by vkruglik-aka** [sqlalchemy/sqlalchemy#13369] [Usage Questions]
Dear all, we're occasionally seeing the otel db.client.connections.usage "used" state UpDownCounter dropping into negative territory for the QueuePool.

I suspect this happens because the opentelemetry python contrib for sqlalchemy may be taking unnecessary or wrong sqlalchemy events into consideration or not taking something else into consideration when updating the counters.

I am asking for help from the sqlalchemy SMEs and/or community to review the otel integration code (pretty small) and see if you can spot the issue(s) with their approach.

The metric is documented here: https://github.com/open-telemetry/semantic-conventions/blob/v1.24.0/docs/database/database-metrics.md#metric-dbclientconnectionsusage

They track two connection states:
1. "used" - connection is checked out from connection pool, presumed to be in-use
2. "idle" - connection is not in-use, presumed to be in the connection pool.

The sqlalchemy otel integration in question is here: https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py

At the time of this writing, the implementation registers for the following sqlalchemy events to update the "used" and "idle" UpDownCounters. I am annotating each registered events with the actions that their event handler takes:

1. self._register_event_listener(engine, "connect", self._pool_connect)
   * self._add_idle_to_connection_usage(1)
2. self._register_event_listener(engine, "close", self._pool_close)
   * self._add_idle_to_connection_usage(-1)
   * Usage of the "close" event for updating the "idle" count seems suspect to me because it's emitted before close is called and close may fail if connection was already closed. So, this could be double-counting and undeflowing the "idle" count? See https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.close.
3. self._register_event_listener(engine, "checkin", self._pool_checkin)
   * self._add_used_to_connection_usage(-1)
   * self._add_idle_to_connection_usage(1)
5. self._register_event_listener(engine, "checkout", self._pool_checkout)
   * self._add_idle_to_connection_usage(-1)
   * self._add_used_to_connection_usage(1)

**Comment by CaselIT** [comment]
Hi,

I think there are some missing events and missing assumptions here.
Try the following script
```py
from sqlalchemy import event
from sqlalchemy.engine.create import create_engine

engine = create_engine("postgresql://scott:tiger@localhost/test")


@event.listens_for(engine, "connect")
def pool_connect(dbapi_connection, connection_record):
    print("pool_connect", dbapi_connection)


@event.listens_for(engine, "close")
def pool_close(dbapi_connection, connection_record):
    print("pool_close", dbapi_connection)


@event.listens_for(engine, "checkin")
def pool_checkin(dbapi_connection, connection_record):
    print("pool_checkin", dbapi_connection)


@event.listens_for(engine, "checkout")
def pool_checkout(dbapi_connection, connection_record, connection_proxy):
    print("pool_checkout", dbapi_connection)


@event.listens_for(engine, "detach")
def pool_detach(dbapi_connection, connection_record):
    print("pool_detach", dbapi_connection)


@event.listens_for(engine, "close_detached")
def pool_close_detached(dbapi_connection):
    print("pool_close_detached", dbapi_connection)


print(1)
conn = engine.connect()
conn.close()
print(2)
with engine.connect() as conn:
    pass
print(3)
engine.dispose()
print(4)
conn = engine.connect()
conn.detach()
conn.close()
print(5)
conn = engine.connect()
conn.invalidate()
print(6)
```

I think you need to at least:
- listen to detach to remove one checked out connection from the pool, don't add to the idle connection
- edit the checkin event to only increment idle conn if there is still a dbapi

**Comment by vkruglik-aka** [comment]
Thank you for following up @CaselIT ! What do you think of the following assessment?

Usage of the "close" event for updating the "idle" count in opentelemetry also seems suspect to me because it's emitted before close is called on the underlying connection and close may fail if connection was already closed. So, this could be double-counting and undeflowing the "idle" count? See https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.close.

**Reply by CaselIT** [reply]
note that what may fail is the closure of the dbapi connection, not the fact that the close happens from sqlalchemy point of view.

I'm not really sure how you can underflow the idle count, I would expect it to increment from what you are describing, since you would be counting multiple times the same idle connection as being idle.

I think one issue may be that you cannot know if the close is happening because of a
- `discpose()` -> the connection was idle, so -1 is appropriate here
- `invalidate()` -> the connection was checked out, so you would need to do a -1 on the checked out connections, not on the idle ones. 

you could listen to the invalidate event for the second case, and for the fist case only consider [valid connections](https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.PoolProxiedConnection.is_valid) for the fist case
